File: backend/src/main/java/com/civicpulse/model/entity/Complaint.java
package com.civicpulse.model.entity;

import com.civicpulse.model.enums.ComplaintStatus;
import com.civicpulse.model.enums.PriorityLevel;
import com.civicpulse.model.enums.SeverityLevel;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "complaints")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Complaint {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private UUID id;

    @Column(name = "complaint_number", nullable = false, unique = true)
    private String complaintNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "citizen_id", nullable = false)
    private User citizen;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private ComplaintCategory category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "department_id")
    private Department department;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "zone_id")
    private CityZone zone;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private ComplaintStatus status = ComplaintStatus.REPORT_SUBMITTED;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    private PriorityLevel priority = PriorityLevel.MEDIUM;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    private SeverityLevel severity = SeverityLevel.MEDIUM;

    @Column(name = "affected_count")
    @Builder.Default
    private Integer affectedCount = 1;

    @Column(columnDefinition = "TEXT")
    private String address;

    private String landmark;
    private String city;
    private String pincode;

    @Column(name = "sla_deadline")
    private Instant slaDeadline;

    @Column(name = "sla_breached")
    @Builder.Default
    private Boolean slaBreached = false;

    @Column(name = "is_emergency")
    @Builder.Default
    private Boolean isEmergency = false;

    @Column(name = "is_anonymous")
    @Builder.Default
    private Boolean isAnonymous = false;

    @Column(name = "resolved_at")
    private Instant resolvedAt;

    @Column(name = "closed_at")
    private Instant closedAt;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private Instant updatedAt;

    @Column(name = "deleted_at")
    private Instant deletedAt;
}

File: backend/src/models/Complaint.js
import mongoose from 'mongoose';

const complaintSchema = new mongoose.Schema({
  complaintNumber: { type: String, required: true, unique: true },
  citizen: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  category: { type: mongoose.Schema.Types.ObjectId, ref: 'ComplaintCategory', required: true },
  department: { type: mongoose.Schema.Types.ObjectId, ref: 'Department' },
  zone: { type: mongoose.Schema.Types.ObjectId, ref: 'CityZone' },
  title: { type: String, required: true },
  description: { type: String, required: true },
  status: { 
    type: String, 
    enum: ['REPORT_SUBMITTED', 'IN_REVIEW', 'ASSIGNED', 'IN_PROGRESS', 'RESOLVED', 'CLOSED', 'REJECTED'], 
    default: 'REPORT_SUBMITTED' 
  },
  priority: { type: String, enum: ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL'], default: 'MEDIUM' },
  severity: { type: String, enum: ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL'], default: 'MEDIUM' },
  affectedCount: { type: Number, default: 1 },
  address: { type: String },
  landmark: { type: String },
  city: { type: String },
  pincode: { type: String },
  // GeoJSON for exact location
  location: {
    type: { type: String, enum: ['Point'], required: false },
    coordinates: { type: [Number], required: false } // [longitude, latitude]
  },
  slaDeadline: { type: Date },
  slaBreached: { type: Boolean, default: false },
  isEmergency: { type: Boolean, default: false },
  isAnonymous: { type: Boolean, default: false },
  resolvedAt: { type: Date },
  closedAt: { type: Date },
  deletedAt: { type: Date }
}, {
  timestamps: true
});

complaintSchema.index({ location: '2dsphere' });

const Complaint = mongoose.model('Complaint', complaintSchema);
export default Complaint;
